File: lib/sweeper.ts
/**
 * Idle-room sweeper.
 *
 * Closing a browser tab doesn't trigger a `leaveRoom` action — only a
 * `markDisconnected` call from the SSE close handler. In `waiting`
 * phase the player stays in the array; in any phase, the room stays
 * alive even if everyone walks away. After a server restart, those
 * rooms come back from disk and never clean themselves up. This
 * sweeper scans the `rooms` table on a fixed interval and evicts any
 * code whose `updated_at` is past a phase-aware TTL.
 *
 * "Don't sweep something that's actively held": if the in-memory Game
 * still has SSE subscribers, we leave it alone — those clients'
 * mutations will bump `updated_at` shortly anyway.
 */

import { deleteRoom, listStaleRoomCodes } from './db';
import { gameStore } from './game-store';

const WAITING_TTL_MS = 60 * 60 * 1_000;        // 1 hour: lobbies, scoreboards
const ACTIVE_TTL_MS = 6 * 60 * 60 * 1_000;     // 6 hours: actual in-flight games
const SWEEP_INTERVAL_MS = 5 * 60 * 1_000;      // every 5 minutes

declare global {
  // eslint-disable-next-line no-var
  var __phantom_cipher_sweeper__: NodeJS.Timeout | undefined;
}

function sweepOnce(): void {
  let codes: string[];
  try {
    codes = listStaleRoomCodes(Date.now(), WAITING_TTL_MS, ACTIVE_TTL_MS);
  } catch (err) {
    console.error('[sweeper] listStaleRoomCodes failed:', err);
    return;
  }
  if (codes.length === 0) return;

  let removed = 0;
  for (const code of codes) {
    const game = gameStore.get(code);
    // Someone is connected right now — let them keep playing. Their
    // next action will refresh updated_at and remove the row from the
    // stale set on the next pass.
    if (game && game.subscriberCount() > 0) continue;
    gameStore.delete(code);
    deleteRoom(code);
    removed++;
  }
  if (removed > 0) {
    console.log(`[sweeper] removed ${removed} idle room(s) (${codes.length - removed} skipped: still connected)`);
  }
}

/**
 * Idempotent: a globalThis flag prevents Next.js dev HMR from stacking
 * up duplicate intervals across module reloads. Sweeps once
 * immediately to clear anything left behind by a previous process.
 */
export function startSweeper(): void {
  if (globalThis.__phantom_cipher_sweeper__) return;
  sweepOnce();
  globalThis.__phantom_cipher_sweeper__ = setInterval(sweepOnce, SWEEP_INTERVAL_MS);
  // Don't keep the event loop alive just for this — if everything
  // else exits, the process should still shut down cleanly.
  globalThis.__phantom_cipher_sweeper__.unref?.();
}

File: backend/src/sweeper.rs
//! 闲置房间清扫器。
//!
//! 关页面不会触发显式 leave，只会 markDisconnected——所以光靠玩家
//! 行为没办法把"开了房没人来"或者"重启之后所有人都没回来"的
//! 僵尸房间清掉。这里按相位分两档 TTL 兜底：
//!   - waiting / ended：1 小时（等待大厅 + 终局画面）
//!   - 进行中：6 小时（一局打两小时也算正常）
//!
//! 还有订阅者就跳过。启动时立即扫一次，之后每 5 分钟一轮。

use crate::db;
use crate::store::Store;
use std::sync::Arc;
use std::time::Duration;
use tokio::time;

const WAITING_TTL_MS: i64 = 60 * 60 * 1_000;
const ACTIVE_TTL_MS: i64 = 6 * 60 * 60 * 1_000;
const SWEEP_INTERVAL: Duration = Duration::from_secs(5 * 60);

pub fn spawn(store: Arc<Store>) {
    tokio::spawn(async move {
        // Immediate sweep on boot, then every interval.
        sweep_once(&store);
        let mut ticker = time::interval(SWEEP_INTERVAL);
        ticker.tick().await; // first tick fires immediately; consume.
        loop {
            ticker.tick().await;
            sweep_once(&store);
        }
    });
}

fn sweep_once(store: &Store) {
    let now = now_ms();
    let candidates = db::list_stale_room_codes(&store.db, now, WAITING_TTL_MS, ACTIVE_TTL_MS);
    if candidates.is_empty() {
        return;
    }
    let mut removed = 0;
    let mut skipped = 0;
    for code in &candidates {
        if let Some(room) = store.get(code) {
            if room.subscriber_count() > 0 {
                skipped += 1;
                continue;
            }
        }
        store.remove(code);
        db::delete_room(&store.db, code);
        removed += 1;
    }
    if removed > 0 {
        tracing::info!(
            removed,
            skipped,
            "sweeper: removed {removed} idle room(s) (skipped {skipped})"
        );
    }
}

fn now_ms() -> i64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}
